src/foreman_mcp_server/middleware/auth.py

```diff
@@ -1,21 +1,44 @@
 import apypie
 from fastmcp.server.middleware import Middleware, MiddlewareContext
 
+
 class AuthMiddleware(Middleware):
     def __init__(self, foreman_url: str):
+        # TODO: user_map can grow indefinitely, consider using a more efficient structure or cleanup mechanism
+        self.user_map = {}
         self.foreman_url = foreman_url
 
-    async def on_call_tool(self, context: MiddlewareContext, call_next):
+    async def on_request(self, context: MiddlewareContext, call_next):
         # Add the foreman_api to the context for use in tools
         if context.fastmcp_context:
-            foreman_username = context.fastmcp_context.request_context.request.headers.get("foreman_username")
-            foreman_password = context.fastmcp_context.request_context.request.headers.get("foreman_token")
-            foreman_api = apypie.ForemanApi(
-                uri=self.foreman_url,
-                username=foreman_username,
-                password=foreman_password,
-                verify_ssl=False,
+            foreman_username = (
+                context.fastmcp_context.request_context.request.headers.get(
+                    "foreman_username"
+                )
+            )
+            foreman_token = context.fastmcp_context.request_context.request.headers.get(
+                "foreman_token"
+            )
+            if not foreman_username or not foreman_token:
+                raise RuntimeError(
+                    "Foreman username and token must be provided in headers."
+                )
+            session_id = context.fastmcp_context.request_context.request.headers.get(
+                "mcp-session-id"
             )
-            setattr(context.fastmcp_context, 'foreman_api', foreman_api)
+            if (
+                not self.user_map.get(foreman_username)
+                or session_id not in self.user_map[foreman_username]
```

Comment:
This is a little bit easier on the eyes
```suggestion
                session_id not in self.user_map.get(foreman_username, {})
```

Reply:
Done.